metricbeat/module/kafka/broker.go

```diff
@@ -290,6 +295,68 @@ func (b *Broker) FetchPartitionOffsetFromTheLeader(topic string, partitionID int
 	return offset, nil
 }
 
+// FetchPartitionOffsetsForTopics fetches offsets of all partitions from its leaders.
+func (b *Broker) FetchPartitionOffsetsForTopics(topics []*sarama.TopicMetadata, time int64) map[string]map[int32]*PartitionOffsets {
+	leaderTopicPartition := map[int32]map[string]int32{}
+	leaderBrokers := map[int32]*sarama.Broker{}
+
+	for _, topic := range topics {
+		for _, partition := range topic.Partitions {
+			broker, err := b.client.Leader(topic.Name, partition.ID)
```

Comment:
Using the method `b.client.Leader(topic, partition)` will always return the most actual leader (there might be a metadata update in background, right?).

The method `b.client.Brokers()` returns brokers without opening connections to them. To establish a connection, I would need a configuration structure: `broker.Open(conf *Config)`. `Leader()` handles it on its own.

Regarding `Close()`, I think you're right.

Reply:
There can still be a problem with this approach, here we are calling `Leader` for each topic and partition, this may open too many connections to brokers, and we may be leaking connections because we only keep track of one connection per broker in `leaderBrokers`.

> Using the method `b.client.Leader(topic, partition)` will always return the most actual leader (there might be a metadata update in background, right?).

This is right, but between this moment and the moment we make the offsets request there can still be some metadata change, if we want to solve this for good (not sure if it worths it) we would need to handle leadership errors (second option in https://github.com/elastic/beats/issues/13380).

Reply:
Answered below.